### ISSUE TITLE
No way to bind attestations in attestation bundle together

### ISSUE THREAD

**Issue by colek42** [in-toto/attestation#136]
In Witness, we use an attestation_collection.  Our current implementation nests all of the attestations into a single object, and signs them together.  This allows us to bind attestations together if they are part of the same process.  For example, if we have an SLSA and a runtime trace attestation, the user should have a way to verify that they happened as a part of the same invocation.

One way forward would be to create an "attestation_collection" type that contains the hashes of the attestations created during the invocation.

The corresponding issue in Witness is here: https://github.com/testifysec/witness/issues/240

**Comment by colek42** [comment]
We want to (eventually) be able to use different envelope types.  Also, the "attestaion_collection"should include subjects for all of the attestations that we are binding together, this is important as it allows us to index the attestations in archivista.

**Comment by colek42** [comment]
We would want to calculate the secure hash over the statements.  We do not want to be bound to a specific envelope type. 

Also, the collection should expose all subjects for its referencing statements.

@mikhailswift probably has some thoughts as well

**Comment by MarkLodato** [comment]
Sorry, I'm not sure I understand the use case. Could you expand on your example, preferably giving example data? Like, do you want to just refer to one attestation from another, or sign multiple statements as a unit (in a single envelope), or create a new (signed) attestation that "bundles" multiple existing (signed) attestations?

**Comment by colek42** [comment]
@MarkLodato go to https://judge.testifysec.io/ and type in `https://github.com/testifysec/go-witness`  each of the results is an attestation collection.  Those attestations happened during the same invocation of Witness.

**Comment by MarkLodato** [comment]
More potential use cases for SLSA:

- https://github.com/slsa-framework/slsa/issues/620: Perhaps we want to suggest creating multiple provenance predicates representing different levels of abstraction?
- @lehors brought up the idea of dropping `resolvedDependencies` in favor of using a standard SBOM format (CycloneDX or SPDX), which would similarly want to bind the two predicates together.

**Comment by TomHennen** [comment]
We discussed inclusion of https://github.com/in-toto/attestation/issues/136 at today's maintainers meeting. While we understand the use case combining a lot of predicates into one statement can be an anti-pattern in that it prevents users from filtering out the predicate types they're not interested in if they want to only forward some of them (e.g. they're calling an API that has concerns about sending it huge amounts of data).

So we've decided it's too risky to include in v1, but are happy to discuss post-v1.

One idea is to have a predicate type that is a list of references to existing DSSEs.  That way the individuals statements could be filtered if needed, but folks still have a way of knowing all the evidence that is bundled together.

**Comment by TomHennen** [comment]
Oh note that it may be entirely possible to handle this just by defining a new predicateType, which won't block on versioning.

**Comment by marcelamelara** [comment]
Was thinking about this problem today. Could the [SCAI](https://github.com/in-toto/attestation/blob/main/spec/predicates/scai.md) predicate be used to address this? The predicate needs to be updated to use the v1.0 spec, but I'm thinking it could be used to implement an attestation collection like this (using the example above). Thoughts?

Pros: Existing predicate type that already has support for artifact attributes and evidence
Cons: Adds a level of indirection, and best way to set the SCAI `attribute` field is still an open question

```
{
    // Standard attestation fields
    "_type": "https://in-toto.io/Statement/v1",
    "subject": [{ ... A ... }],
        
    "predicateType": "https://in-toto.io/scai/attribute-report/v0.2?draft",
    "predicate": {
        "attributes": [{
            "attribute": "https://witness.dev/attestations/github/v0.1",
            "evidence": {               // this will be a ResourceDescriptor in v0.2
                "digest": { "sha256": "abcdabcde..." },
                "content": "<Base64(serialized DSSE envelope)>",
                "mediaType": "application/x.dsse+json"
            }
        },
        {
            "attribute": "https://witness.dev/attestations/material/v0.1",
            "evidence":  { 
                "digest": { "sha256": "01234567..." },
                "content": "<Base64(serialized DSSE envelope)>",
                "mediaType": "application/x.dsse+json"
            }
        }]
    }
}
```

**Comment by TomHennen** [comment]
Oh, I should note that you could also do that with the SCAI predicate, you just wouldn't include the content directly.

**Comment by marcelamelara** [comment]
Yes, completely agree that including the `content` of attestations might be too large in practice. In your example, I guess you're assuming that the evidenceCollection attestation would be included in the same bundle as the DSSEs in the collection? Otherwise, I suppose the `evidence` field for each DSSE could include the `uri` or `downloadLocation`.

**Comment by marcelamelara** [comment]
@colek42 @mikhailswift @MarkLodato I think that the existing SCAI predicate may be able to address the use cases listed in this issue. PTAL at #170 . Would appreciate any input!
